Crash when async functions return non-PromiseLike thenables.

```ts
export interface MyThenable {
    then(): void;
}

export async function foo(): MyThenable {
    return {
        then() {
        }
    };
}
```

Compile this and you'll get the following error.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered, repanicked]
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x84ca6e]

goroutine 313 [running]:
sync.(*WaitGroup).Go.func1.1()
        sync/waitgroup.go:251 +0x45
panic({0xd7e040?, 0x1b9f240?})
        runtime/panic.go:860 +0x13a
github.com/microsoft/typescript-go/internal/checker.(*Checker).unwrapAwaitedType(0x242b39e12008?, 0x242b37e24608?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:30734 +0xe
github.com/microsoft/typescript-go/internal/checker.(*Checker).createPromiseLikeType(0x242b39e12008, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:19979 +0x4e
github.com/microsoft/typescript-go/internal/checker.(*Checker).getContextualTypeForReturnExpression(0x242b39e12008, 0x242b3a8c7470?, 0x37d11008?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:28941 +0x19a
github.com/microsoft/typescript-go/internal/checker.(*Checker).getContextualType(0x242b39e12008, 0x242b37d7c730?, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:28656 +0x385
github.com/microsoft/typescript-go/internal/checker.(*Checker).pushCachedContextualType(...)
        github.com/microsoft/typescript-go/internal/checker/checker.go:30166
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkObjectLiteral(0x242b39e12008, 0x242b37d7c730, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:12860 +0x1f7
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkExpressionWorker(0x242b39e12008, 0x242b37d7c730, 0x8067bf?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:7596 +0x277
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkExpressionEx(0x242b39e12008, 0x242b37d7c730, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:7395 +0x1dd
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkExpressionCachedEx(0x242b39e12008, 0x242b37d7c730, 0x48a6c6?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:7361 +0xd7
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkExpressionCached(...)
        github.com/microsoft/typescript-go/internal/checker/checker.go:7345
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkReturnStatement(0x242b39e12008, 0x242b37e26e10)
        github.com/microsoft/typescript-go/internal/checker/checker.go:4033 +0x19e
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x242b39e12008, 0x242b37e26e10)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2295 +0x354
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x242b39e12008, 0x242b37e24608?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2196 +0x65
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElements(0x242b39e12008, {0x242b37e72258, 0x1, 0x7b7155?})
        github.com/microsoft/typescript-go/internal/checker/checker.go:2186 +0x2d
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkBlock(0x242b39e12008, 0x242b37e78360)
        github.com/microsoft/typescript-go/internal/checker/checker.go:3725 +0xfc
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x242b39e12008, 0x242b37e78360)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2275 +0x3c5
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x242b39e12008, 0x242b3885ad80?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2196 +0x65
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkFunctionOrMethodDeclaration(0x242b39e12008, 0x242b37e24608)
        github.com/microsoft/typescript-go/internal/checker/checker.go:3372 +0x14e
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkFunctionDeclaration(0x242b39e12008, 0x242b37e24608)
        github.com/microsoft/typescript-go/internal/checker/checker.go:3342 +0x1d
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x242b39e12008, 0x242b37e24608)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2273 +0x39f
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x242b39e12008, 0x242b39e7e570?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2196 +0x65
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElements(0x242b39e12008, {0x242b37e34e40, 0x2, 0x242b3a8c7cc0?})
        github.com/microsoft/typescript-go/internal/checker/checker.go:2186 +0x2d
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceFile(0x242b39e12008, {0x1167298?, 0x1bfc7a0?}, 0x242b37e54e08, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2154 +0x205
github.com/microsoft/typescript-go/internal/checker.(*Checker).getDiagnostics(0x242b39e12008, {0x1167298?, 0x1bfc7a0?}, 0x242b37e54e08, 0x242b39e12200)
        github.com/microsoft/typescript-go/internal/checker/checker.go:13633 +0x67
github.com/microsoft/typescript-go/internal/checker.(*Checker).GetDiagnostics(...)
        github.com/microsoft/typescript-go/internal/checker/checker.go:13623
github.com/microsoft/typescript-go/internal/compiler.(*Program).getBindAndCheckDiagnosticsWithChecker(0x242b37e20788, {0x1167298, 0x1bfc7a0}, 0x242b39e12008, 0x242b37e54e08)
        github.com/microsoft/typescript-go/internal/compiler/program.go:1306 +0xc5
github.com/microsoft/typescript-go/internal/compiler.(*Program).getSemanticDiagnosticsWithChecker(0x242b37e20788, {0x1167298?, 0x1bfc7a0?}, 0x242b39e4fe18?, 0x242b37e54e08)
        github.com/microsoft/typescript-go/internal/compiler/program.go:1290 +0x2e
github.com/microsoft/typescript-go/internal/compiler.(*Program).collectCheckerDiagnosticsFromFiles.func1(0xd9f8c0?, 0x52, 0x242b37e54e08?)
        github.com/microsoft/typescript-go/internal/compiler/program.go:553 +0x47
github.com/microsoft/typescript-go/internal/compiler.(*checkerPool).forEachCheckerGroupDo.func1()
        github.com/microsoft/typescript-go/internal/compiler/checkerpool.go:159 +0x159
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1()
        github.com/microsoft/typescript-go/internal/core/workgroup.go:40 +0x13
sync.(*WaitGroup).Go.func1()
        sync/waitgroup.go:258 +0x4a
created by sync.(*WaitGroup).Go in goroutine 1
        sync/waitgroup.go:238 +0x73
```

--- a/internal/checker/checker.go
+++ b/internal/checker/checker.go
@@ -28938,6 +28938,9 @@ func (c *Checker) getContextualTypeForReturnExpression(node *ast.Node, contextFl
 			if functionFlags&ast.FunctionFlagsAsync != 0 {
 				// Get the awaited type without the `Awaited<T>` alias
 				contextualAwaitedType := c.mapType(contextualReturnType, c.getAwaitedTypeNoAlias)
+				if contextualAwaitedType == nil {
+					return nil
+				}
 				return c.getUnionType([]*Type{contextualAwaitedType, c.createPromiseLikeType(contextualAwaitedType)})
 			}
 			// Regular function or Generator function

--- a/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.errors.txt
+++ b/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.errors.txt
@@ -1,0 +1,20 @@
+asyncFunctionReturnNonPromiseThenable.ts(5,30): error TS1064: The return type of an async function or method must be the global Promise<T> type. Did you mean to write 'Promise<void>'?
+asyncFunctionReturnNonPromiseThenable.ts(6,5): error TS1058: The return type of an async function must either be a valid promise or must not contain a callable 'then' member.
+
+
+==== asyncFunctionReturnNonPromiseThenable.ts (2 errors) ====
+    export interface MyThenable {
+        then(): void;
+    }
+    
+    export async function foo(): MyThenable {
+                                 ~~~~~~~~~~
+!!! error TS1064: The return type of an async function or method must be the global Promise<T> type. Did you mean to write 'Promise<void>'?
+        return {
+        ~~~~~~
+!!! error TS1058: The return type of an async function must either be a valid promise or must not contain a callable 'then' member.
+            then() {
+            }
+        };
+    }
+    

--- a/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.js
+++ b/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.js
@@ -1,0 +1,22 @@
+//// [tests/cases/compiler/asyncFunctionReturnNonPromiseThenable.ts] ////
+
+//// [asyncFunctionReturnNonPromiseThenable.ts]
+export interface MyThenable {
+    then(): void;
+}
+
+export async function foo(): MyThenable {
+    return {
+        then() {
+        }
+    };
+}
+
+
+//// [asyncFunctionReturnNonPromiseThenable.js]
+export async function foo() {
+    return {
+        then() {
+        }
+    };
+}

--- a/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.symbols
+++ b/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.symbols
@@ -1,0 +1,21 @@
+//// [tests/cases/compiler/asyncFunctionReturnNonPromiseThenable.ts] ////
+
+=== asyncFunctionReturnNonPromiseThenable.ts ===
+export interface MyThenable {
+>MyThenable : Symbol(MyThenable, Decl(asyncFunctionReturnNonPromiseThenable.ts, 0, 0))
+
+    then(): void;
+>then : Symbol(MyThenable.then, Decl(asyncFunctionReturnNonPromiseThenable.ts, 0, 29))
+}
+
+export async function foo(): MyThenable {
+>foo : Symbol(foo, Decl(asyncFunctionReturnNonPromiseThenable.ts, 2, 1))
+>MyThenable : Symbol(MyThenable, Decl(asyncFunctionReturnNonPromiseThenable.ts, 0, 0))
+
+    return {
+        then() {
+>then : Symbol(then, Decl(asyncFunctionReturnNonPromiseThenable.ts, 5, 12))
+        }
+    };
+}
+

--- a/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.types
+++ b/testdata/baselines/reference/compiler/asyncFunctionReturnNonPromiseThenable.types
@@ -1,0 +1,20 @@
+//// [tests/cases/compiler/asyncFunctionReturnNonPromiseThenable.ts] ////
+
+=== asyncFunctionReturnNonPromiseThenable.ts ===
+export interface MyThenable {
+    then(): void;
+>then : () => void
+}
+
+export async function foo(): MyThenable {
+>foo : () => MyThenable
+
+    return {
+>{        then() {        }    } : { then(): void; }
+
+        then() {
+>then : () => void
+        }
+    };
+}
+

--- a/testdata/tests/cases/compiler/asyncFunctionReturnNonPromiseThenable.ts
+++ b/testdata/tests/cases/compiler/asyncFunctionReturnNonPromiseThenable.ts
@@ -1,0 +1,13 @@
+// @strict: true
+// @target: esnext
+
+export interface MyThenable {
+    then(): void;
+}
+
+export async function foo(): MyThenable {
+    return {
+        then() {
+        }
+    };
+}

